[FLINK-5104] Bipartite graph validation

mushketyk opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed


This PR duplicates some the user documentation for the graph added in https://github.com/apache/flink/pull/2984 but I think this won't be an issue when documentation PR is pushed to the `master`.

mushketyk commented:
I noticed that `Graph` has the following implementation of the `validate` method:

```java
public Boolean validate(GraphValidator<K, VV, EV> validator) throws Exception {
	return validator.validate(this);
}
```

What is the reason for returning `Boolean` instead of `boolean` here (GraphValidator returns `boolean`)?
Shouldn't we change the return value to `boolean`?

greghogan commented on flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/validation/InvalidBipartiteVertexIdsValidator.java:
```diff
@@ -0,0 +1,85 @@
+
+	@Override
+	public boolean validate(BipartiteGraph<KT, KB, VVT, VVB, EV> bipartiteGraph) throws Exception {
+		DataSet<Tuple1<KT>> edgesTopIds = bipartiteGraph.getEdges().map(new GetTopIdsMap<KT, KB, EV>());
```
Can `GetTopIdsMap` be replaced with `.<Tuple1<KT>>project(0)`?

mushketyk commented on flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/validation/InvalidBipartiteVertexIdsValidator.java:
```diff
@@ -0,0 +1,85 @@
+		DataSet<KT> invalidTopIds = invalidIds(bipartiteGraph.getTopVertices(), edgesTopIds);
+		DataSet<KB> invalidBottomIds = invalidIds(bipartiteGraph.getBottomVertices(), edgesBottomIds);
+
+		return invalidTopIds.count() == 0 && invalidBottomIds.count() == 0;
```
Since these two datasets have different types I couldn't figure out how to use accumulators and `RichOutputFormat`  to run `collect()` only once, but I've replaced it with two outer joins and one `count()` call.
Is it a feasible approach?

mushketyk commented:
Hi @greghogan ,

Thank you for your feedback. I've updated the PR accordingly. The only thing that I did differently is that I've replaced projections and two `count()` executions with joins and one `count()` call. What do you think about this?

greghogan commented:
@mushketyk the build is failing due to unused imports (IntelliJ can be configured to automatically remove these).

There are so many ways to implement the validator. Projections would reduce the data to be transmitted and sorted. I like to document when we are performing a different kind of join (here, an anti-join) for the day when these are available in Flink. I don't know how counting in the `FlatJoinFunction` followed by two `DiscardingOutputFormat` compares with the current implementation of a union and count (which is a dummy `OutputFormat`).

We'll need to rebase this PR once FLINK-5311 has been committed.

mushketyk commented:
Sorry for the failing build. I've removed the unused imports so it should be fine now.

mushketyk commented:
@greghogan Thank you for the review.
Can you give me an example of how you document joins?

greghogan commented:
@mushketyk I had meant just with a simple comment noting that with an anti-join we could also remove the FlatJoinFunction.

mushketyk commented:
Rebased this PR on top of `master`.

aljoscha commented:
I'm closing this as "Abandoned", since there is no more activity and the code base has moved on quite a bit. Please re-open this if you feel otherwise and work should continue.
